### ISSUE TITLE
chapter 4) run될때마다 초기화?, 편향을 왜 주는지?

### ISSUE THREAD

**Issue by younseun** [gcpugkr/study_201807#3]
layers.py의 내용

import tensorflow as tf

def weight_variable(shape):
    initial = tf.truncated_normal(shape, stddev=0.1)
    return tf.Variable(initial)

def bias_variable(shape):
    initial = tf.constant(0.1, shape=shape)
    return tf.Variable(initial)

def conv2d(x, W):
    return tf.nn.conv2d(x, W, strides=[1, 1, 1, 1], padding='SAME')

def max_pool_2x2(x):
    return tf.nn.max_pool(x, ksize=[1, 2, 2, 1],
                          strides=[1, 2, 2, 1], padding='SAME')

def conv_layer(input, shape):
    W = weight_variable(shape)
    b = bias_variable([shape[3]])
    return tf.nn.relu(conv2d(input, W) + b)

def full_layer(input, size):
    in_size = int(input.get_shape()[1])
    W = weight_variable([in_size, size])
    b = bias_variable([size])
    return tf.matmul(input, W) + b

-----------------------------------------------------------------
그래프 생성 Code
-----------------------------------------------------------------

cifar = CifarDataManager()

x = tf.placeholder(tf.float32, shape=[None, 32, 32, 3])
y = tf.placeholder(tf.float32, shape=[None, 10])
keep_prob = tf.placeholder(tf.float32)

x_placeholder = x
conv1 = conv_layer(x, shape=[5,5,3,32])
conv1_pool = max_pool_2x2(conv1)
conv2 = conv_layer(conv1_pool, shape=[5,5,32,64])
conv2_pool = max_pool_2x2(conv2)

conv2_plat = tf.reshape(conv2_pool, [-1, 8*8*64])
full_1 = tf.nn.relu(full_layer(conv2_plat, 1024))
full_drop = tf.nn.dropout(full_1, keep_prob)

y_conv = full_layer(full_drop, 10)

cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(labels=y, logits=y_conv))
train_step = tf.train.AdamOptimizer(1e-3).minimize(cross_entropy)

correct_predict = tf.equal(tf.argmax(y_conv, 1), tf.argmax(y,1))
accuracy =  tf.reduce_mean(tf.cast(correct_predict, tf.float32))

-----------------------------------------------------------------
-- 그래프 실행
-----------------------------------------------------------------
STEPS = 1000
X = cifar.test.images.reshape(10, 1000, 32, 32, 3)
Y = cifar.test.labels.reshape(10, 1000, 10)

def test(sess):
    X = cifar.test.images.reshape(10, 1000, 32, 32, 3)
    Y = cifar.test.labels.reshape(10, 1000, 10)
    acc = np.mean([sess.run(accuracy, feed_dict={x: X[i], y_: Y[i], keep_prob: 1.0})
                   for i in range(10)])
    print("Accuracy: {:.4}%".format(acc * 100))

with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    for i in range(STEPS):
        batch = cifar.train.next_batch(BATCH_SIZE)
        outs = sess.run(train_step, feed_dict={x:batch[0],y:batch[1], keep_prob: 0.5})
        if (i+1) % 100 == 0:   # 평가 실시
            step_outs_accuracy = np.mean([sess.run([accuracy], feed_dict={x: X[j], y: Y[j], keep_prob: 1.0}) for j in range(10)])
            print('{} : Accuraty : {}            {} '.format(i+1, step_outs_accuracy, datetime.datetime.now()))
    print('Completed')

-----------------------------------------------------------------
-- 질문 내용
-----------------------------------------------------------------
1. 매번 초기화?
run으로 그래프를 실행할때마다.
conv_layer를 수행하는데, 
conv_layer함수에서는 weight_variable함수를 Call하고,
weight_variable함수에서는 사용할 weight를 초기화 한다.
 Training을 할 수록 weight의 variable의 값을 변경해 나가는것으로 이해하는데, run을 할때마다 초기화 함수를 타는것을 모르겠다.

2. bias_variable의 0.1편향
합성곱 계층 적용시 "tf.nn.relu(conv2d(input, W) + b)"를 적용하는데, 왜 모든 weight에 동일한 편향값을 주는지 모르겠다.
편향값을 주지 않으면 accuracy가 높고, 편향값을 줄수록 accuracy가 줄어든다.

**Comment by Ooooooog** [comment]
1. 말씀하신대로 weight는 최초에 초기화 할때만 초기화되고, 그뒤로는 초기화되지 않습니다. 
weight_variable는 처음에 레이어를 구성할때 해당 레이어에 사용될 weight를 초기화하는 데 쓰이고 있습니다.

2.
https://www.quora.com/What-is-bias-in-artificial-neural-network 
bias는 weight와 별개의 모델 파라미터입니다. 
linear regression에서 그래프식을 정할 때 x*w이 0이 되는 것을 방지하지 위해 상수값 bias를 더해주어 x*w+b가 된 것으로 알고있습니다.
이과 비슷한 원리로 dnn의 경우 각 레이어마다 1개의 bias값이 주어져서 계산식 x*W+b에 사용됩니다.
(여기서 x와 W는 모두 다차원 백터이지만, b는 상수값입니다)
그러므로 bias는 weight와 무관하게 계산식에 사용되는 일종의 상수값이라고 보시면 될것 같습니다.
